operator/pkg/bgpv2/cluster.go

```diff
@@ -0,0 +1,229 @@
+	var res []v2alpha1.CiliumBGPNodeInstance
+
+	for _, clusterBGPInstance := range clusterBGPInstances {
+		nodeBGPInstance := v2alpha1.CiliumBGPNodeInstance{
+			Name:     clusterBGPInstance.Name,
+			LocalASN: clusterBGPInstance.LocalASN,
+		}
+
+		// find BGPResourceManager global override for this instance
+		var override v2alpha1.CiliumBGPNodeConfigInstanceOverride
+		for _, overrideBGPInstance := range overrideBGPInstances {
+			if overrideBGPInstance.Name == clusterBGPInstance.Name {
+				nodeBGPInstance.RouterID = overrideBGPInstance.RouterID
+				nodeBGPInstance.LocalPort = overrideBGPInstance.LocalPort
+				override = overrideBGPInstance
+				break
+			}
+		}
+
+		for _, clusterBGPInstancePeer := range clusterBGPInstance.Peers {
+			nodePeer := v2alpha1.CiliumBGPNodePeer{
+				Name:          clusterBGPInstancePeer.Name,
+				PeerAddress:   clusterBGPInstancePeer.PeerAddress,
+				PeerASN:       clusterBGPInstancePeer.PeerASN,
+				PeerConfigRef: clusterBGPInstancePeer.PeerConfigRef,
+			}
+
+			// find BGPResourceManager Peer override for this instance
+			for _, overrideBGPPeer := range override.Peers {
+				if overrideBGPPeer.Name == clusterBGPInstancePeer.Name {
+					nodePeer.LocalAddress = overrideBGPPeer.LocalAddress
+					break
+				}
+			}
+
+			nodeBGPInstance.Peers = append(nodeBGPInstance.Peers, nodePeer)
+		}
+
+		res = append(res, nodeBGPInstance)
+	}
+	return res
+}
+
+// getMatchingNodes returns a map of node names that match the given cluster config's node selector.
+func (b *BGPResourceManager) getMatchingNodes(nodeSelector *slim_meta_v1.LabelSelector, configName string) (map[string]struct{}, error) {
+	labelSelector, err := slim_meta_v1.LabelSelectorAsSelector(nodeSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	// find nodes that match the cluster config's node selector
+	matchingNodes := make(map[string]struct{})
+
+	for _, n := range b.ciliumNodeStore.List() {
+		// nil node selector means match all nodes
+		if nodeSelector == nil || labelSelector.Matches(slim_labels.Set(n.Labels)) {
+			err := b.validNodeSelection(n, configName)
+			if err != nil {
+				b.logger.WithError(err).Errorf("skipping node %s", n.Name)
+				continue
+			}
+			matchingNodes[n.Name] = struct{}{}
+		}
+	}
+
+	return matchingNodes, nil
+}
+
+// deleteStaleNodeConfigs deletes node configs that are not in the expected list for given cluster.
+func (b *BGPResourceManager) deleteStaleNodeConfigs(ctx context.Context, expectedNodes map[string]struct{}, clusterRef string) error {
+	var err error
+	for _, existingNode := range b.nodeConfigStore.List() {
+		_, exists := expectedNodes[existingNode.Name]
+		if exists || !IsOwner(existingNode.GetOwnerReferences(), clusterRef) {
+			continue
+		}
+
+		dErr := b.nodeConfigClient.Delete(ctx, existingNode.Name, meta_v1.DeleteOptions{})
```

Comment:
Is there any chance that this could race with another update occuring concurrently? ex. if a Create() on node config is issued just as this is about to run, this might accidentally remove the new object?

Reply:
Reconciliation code is run by single go routine. If there are concurrent updates triggering reconciliation, they are queued [here](https://github.com/cilium/cilium/pull/28846/files#diff-323b6adce32f208c35e844e9b88e97e4b4c7c3340edd0b456918f5106150eae3R278). 

Store are modified asynchronously - there can be a create which is inflight and not yet sync'd with the store. In which case, it will not be part of the list and we basically ignore that nodeConfig object from this logic. 

I think there might be a condition ( not sure how likely ) where there might be stale nodeconfig objects left over. For eg

1. User create ClusterConfig which matches 10 nodes.
2. Reconcile happens - 10 nodeconfig objects are create.
3. **Problem :  local store is not yet sync'd with API server for some reason.**
4. User updates ClusterConfig to match 6 nodes.
5. Reconcile happens - 6 nodeconfig objects are upserted. 
6. Deletion of stale 4 nodeobjects does not happen since local store does not have any nodeconfigobjects yet.
7. local store gets finally syncs with API server - but it has 10 nodeconfig objects.

So from user prospective, there should be 6 nodeconfig objects based on latest cluster config. But we will have 10 instead. I will try to add a test case to see if it can be triggered. 

In the meantime, I wonder if I am overly cautious about such scenario. Workaround for such issue would be removing and readding cluster config CRD - which will re-trigger reconciliation. Or have a lazy timer based reconciliation which checks that everything is healthy and expected.

Is this something which is blocking for the PR or can be subsequently addressed ?

Reply:
Hi @tommyp1ckles  not sure if you had a chance to review my comment. It is also inline with `rabbit hole of worst case scenarios`, If you're okay we can leave things as is for now and tackle this as a follow up.

Reply:
Opened https://github.com/cilium/cilium/issues/30320 to track this issue.